### DISCUSSION TITLE
load OEM file

### DISCUSSION THREAD

**Discussion by mikezang** [skyfielders/python-skyfield#1010] [Ideas]
Can you provide a new API to load OEM file? So that we can get what we get from TLE file.

**Comment by brandon-rhodes** [comment]
Could you provide an example of an OEM file, that illustrates the format? Or a link to a site providing OEM files?

**Comment by mikezang** [comment]
Here you are!
[CSS_OEM_20241009001030_0001.txt](https://github.com/user-attachments/files/17352023/CSS_OEM_20241009001030_0001.txt)

**Reply by brandon-rhodes** [reply]
Thanks for the example! Having looked over the file, I can see documentation for it (among other places) here:

https://public.ccsds.org/Pubs/502x0b3e1.pdf

I am not, alas, aware of any Skyfield contributors who are currently planning to sit down and implement OEM support in the near term. Everyone seems to be busy using OMM instead.

**Reply by mikezang** [reply]
Is it difficult to use OEM file instead of TLE data?

**Reply by brandon-rhodes** [reply]
You can look over chapter "5 ORBIT EPHEMERIS MESSAGE" of the PDF to see the details. It looks like it would take a few days of work, perhaps, for someone to implement? Parsing the file would, happily, be fairly simple, as the format is easy to read and understand. The work involved would be: implementing any new reference frames that OEM files might use, like the EME2000 reference frame used in your example file `CSS_OEM_20241009001030_0001.txt`; implementing the interpolation methods that OEM files can request through the `INTERPOLATION` and `INTERPOLATION_DEGREE` keywords; and, finally, figuring out what to do with files like this one, that don't specify an `INTERPOLATION`. Would Skyfield have a default option? Or, maybe much better, the person loading the file without an `INTERPOLATION` would have to select an interpolation method themselves.

In any case, my guess is that implementing the reference frame and at least one initial interpolation method would be the biggest tasks.

**Comment by gfireman** [comment]
I second this request.  We use OEM files for planning PACE instrument operations.  TLEs are not updated until after a delta-V, but predicted ephemerides have the expected new orbital parameters built in.

Alas, I have neither the time nor the expertise to implement.

**Comment by gfireman** [comment]
This problem may be more tractable than I'd thought:

- The EME2000 reference frame is identical to J2000, and Skyfield already has an [implementation](https://rhodesmill.org/skyfield/api-framelib.html#skyfield.framelib.ecliptic_J2000_frame).
- We use a cubic Hermite spline for orbit interpolation, and scipy already has an [implementation](https://docs.scipy.org/doc/scipy/reference/generated/scipy.interpolate.CubicHermiteSpline.html).

I'd be willing to give this a shot; what would be the next steps?

**Reply by brandon-rhodes** [reply]
Typically, what's needed is a little script that produces, say, three accurate positions. The little script would load and parse the OEM file; then choose three dates-and-times; and for each of them, go through the steps of producing a position. It would print for each of them something like:

```
Vector desired:  [ 331.61901192  392.18492744 1049.7597825 ]
Vector returned: [ 331.61901192  392.18492744 1049.7597825 ]
```

And they would agree.

I can take it from there. Once the calculation works, it's easy enough for me to package it up into an API in a Skyfield-compliant style. The "desired" value should, I guess, come from some other library that already understands OEM and whose output we can trust.

**Reply by gfireman** [reply]
Cool!  Should the vector be in any particular frame?

**Reply by brandon-rhodes** [reply]
Good question! If the library you're comparing to will support ICRS vectors, then using that, and comparing to the `position.xyz` of the position that you compute with Skyfield is simplest. Skyfield uses ICRS internally, so getting the `.xyz` has no extra cost. Whereas other reference frames require Skyfield to rotate the vector, making the test run a bit slower, and the code do a bit more, than otherwise.

**Reply by gfireman** [reply]
TBH, what I have for comparison is not so much a "library" as a collection of tools that Fred Patt wrote in IDL and were translated into python.  We've been using those for PACE science operations scheduling since before launch.

Give me a minute to learn how to use https://www.space-track.org/documentation#/sgp4; should be able to generate "truth" values.

**Reply by brandon-rhodes** [reply]
No rush — I'll be busy all of next week, so I won't be able to usefully respond until early the following week.

**Reply by gfireman** [reply]
Even easier to use your [sgp4 library](https://pypi.org/project/sgp4/).  Ooh, this is exciting!

What will be the best way to get my demo data and script to you?

**Reply by brandon-rhodes** [reply]
It will probably be easiest to switch over to email? I'll try sending you an email at your work address.

**Comment by faithmckinely-alt** [comment]
Same when there is an updated about this one do mention me

**Comment by faithmckinely-alt** [comment]
It works!
